Support for import/export of instance-level domain blocks/allows for 4.x w/ additional fixes

lenoregilbert opened the pull request:
Updated PR of https://github.com/mastodon/mastodon/pull/15664 w/ Several Additional Fixes
* Adds confirmation page when importing blocked domains 
* Fix for authorization check in domain blocks controller
* Fix for incorrect error strings for domain blocks and e-mail domain blocks
* Merged additional commits for test failures related to changes in the role system since original work was done

Tak committed:
Allow import/export of instance-level domain blocks/allows (#1754)

* Allow import/export of instance-level domain blocks/allows.
Fixes #15095

* Pacify circleci

* Address simple code review feedback

* Add headers to exported CSV

* Extract common import/export functionality to
AdminExportControllerConcern

* Add additional fields to instance-blocked domain export

* Address review feedback

* Split instance domain block/allow import/export into separate pages/controllers

* Address code review feedback

* Pacify DeepSource

* Work around Paperclip::HasAttachmentFile for Rails 6

* Fix deprecated API warning in export tests

* Remove after_commit workaround

(cherry picked from commit 94e98864e39c010635e839fea984f2b4893bef1a)

ClearlyClaire committed:
Add confirmation page when importing blocked domains (#1773)

* Move glitch-soc-specific strings to glitch-soc-specific locale files

* Add confirmation page when importing blocked domains

(cherry picked from commit b91196f4b73fff91997b8077619ae25b6d04a59e)

ClearlyClaire committed:
Fix authorization check in domain blocks controller

(cherry picked from commit 75279377583c6e2aa04cc8d7380c593979630b38)

ClearlyClaire committed:
Fix tests

(cherry picked from commit 9094c2f52c24e1c00b594e7c11cd00e4a07eb431)

ClearlyClaire committed:
Fix domain block export not exporting blocks with only media rejection

(cherry picked from commit 26ff48ee48a5c03a2a4b0bd03fd322529e6bd960)

lenoregilbert committed:
Fix error strings for domain blocks and email-domain blocks

Corrected issue with non-error message used for Mastodon:NotPermittedError in Domain Blocks
Corrected issue Domain Blocks using the Email Domain Blocks message on ActionContoller::ParameterMissing
Corrected issue with Email Domain Blocks using the not_permitted string from "custom emojii's"

lenoregilbert committed:
Ran i18n-tasks normalize to address test failure

lenoregilbert committed:
Removed unused admin.export_domain_blocks.not_permitted string

Removing unused string as indicated by Check i18n

ClearlyClaire commented:
This is essentially porting a PR I approved and merged in glitch-soc, as well as subsequent work I did in glitch-soc, so I obviously approve of the PR on principle.

The test failures seem to be caused by not having ported https://github.com/glitch-soc/mastodon/pull/1804/commits/9094c2f52c24e1c00b594e7c11cd00e4a07eb431 that is required to accommodate the role system change that was introduced after the PR was originally merged in glitch-soc.

ClearlyClaire commented:
now it's https://github.com/glitch-soc/mastodon/commit/26ff48ee48a5c03a2a4b0bd03fd322529e6bd960 that's missing :)

Gargron commented on app/javascript/packs/admin.js:
```diff
@@ -185,6 +185,12 @@ ready(() => {
   const registrationMode = document.getElementById('form_admin_settings_registrations_mode');
   if (registrationMode) onChangeRegistrationMode(registrationMode);
 
+  const checkAllElement = document.querySelector('#batch_checkbox_all');
```
Hm, what are these changes for?

ClearlyClaire replied:
It is because the page may be pre-rendered with a mix of unchecked and checked inputs: everything comes pre-checked except for domains where there are existing follow relationships.

Gargron commented on app/models/admin/import.rb:
```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+# A non-activerecord helper class for csv upload
+class Admin::Import
```
If I understand correctly, this will never clean up uploaded CSV files after they're processed since the model is not persisted but nothing triggers paperclip to destroy the data? The thing is, if you're going to process the file synchronously, you don't even need paperclip, you could just read the tempfile from the upload.

ClearlyClaire replied:
Yeah that sounds like a better way to do it, actually.

enbylenore replied:
@ClearlyClaire Is this something you want to change over at glitch-soc and have me merge into this PR or should I take a look at making the change direectly on mastodon?

ClearlyClaire replied:
This is actually more work than I anticipated, but I should be able to do it. The current code is also a bit more wonky than I thought

ClearlyClaire commented:
I can have a shot at doing it in glitch-soc tomorrow!

ClearlyClaire committed:
Fix various issues with domain block import

- stop using Paperclip for processing domain allow/block imports
- stop leaving temporary files
- better error handling
- assume CSV files are UTF-8-encoded

(cherry picked from commit cad824d8f501b95377e4f0a957e5a00d517a1902)

davew commented:
Hi,
Is there anything about when this is likely to be part of a Mastodon release?
